### PR TITLE
Add tests for source container build

### PR DESCRIPTION
**Pull request by tkdchen**
STONEBLD-2094

* GH workflow action is updated to run the tests.
* Add another requirements-dev for running tests.
* source_build.py is updated a little which is not feature change.
* Major change is adding tests for individual portion of the build
  process and the build process itself.

The source build implementation leverages a few command line tools, e.g.
git, tar and BuildSourceImage (aka. bsi). Tests run on real initialized Git
repositories, but not real tar archives and executable bsi script
intentionally, unless it is necessary.

### CODE REVIEW

File: source-container-build/app/source_build.py [hunk 1]
```diff
@@ -389,8 +391,11 @@ def build_and_push(
     ]
     log.debug("push source image %r", push_cmd)
     run(push_cmd, check=True)
-    with open(digest_file, "r") as f:
-        build_result["image_digest"] = f.read().strip()
+    try:
+        with open(digest_file, "r") as f:
+            build_result["image_digest"] = f.read().strip()
+    finally:
+        os.unlink(digest_file)
```
Comment (chmeliik) [inline source-container-build/app/source_build.py:398]: Could you make this change in a separate commit?

File: source-container-build/app/test_source_build.py [hunk 2]
```diff
@@ -0,0 +1,1154 @@
...
+        packages_dir = os.path.join(cachi2_output_dir, "deps", pkg_mgr_path)
+        os.makedirs(packages_dir, exist_ok=True)
+        if filename.endswith(".tar.gz") or filename.endswith(".tgz"):
+            with tarfile.open(os.path.join(packages_dir, filename), "w:gz"):
+                pass
+        elif filename.endswith(".zip"):
+            with zipfile.ZipFile(os.path.join(packages_dir, filename), "w"):
+                pass
+        else:
+            pass
```
Comment (chmeliik) [inline source-container-build/app/test_source_build.py:118]: It looks like this mocking function ignores files that don't end with .tar.gz, .tgz or .zip?

File: source-container-build/app/test_source_build.py [hunk 3]
```diff
@@ -0,0 +1,1154 @@
...
+
+    def test_gather_npm_deps(self):
+        npm_deps = ["npm/has-1.0.3.tgz", "npm/express-4.18.1.tgz", "npm/bytes-3.1.2.tgz"]
+        self._test_gather_deps_by_package_manager(npm_deps, ".tgz")
+
+    def test_gather_go_deps(self):
+        gomod_deps = [
+            "gomod/pkg/mod/cache/download/gopkg.in/yaml.v2/@v/v2.4.0.zip",
+            "gomod/pkg/mod/cache/download/gopkg.in/yaml.v2/@v/v2.4.0.info",
+            "gomod/pkg/mod/cache/download/github.com/go-logr/logr/@v/v1.2.3.lock",
```
Comment (chmeliik) [inline source-container-build/app/test_source_build.py:760]: I think the `create_fake_dep_packages` function will ignore the `.lock` file, so adding it here doesn't achieve anything.

I think `create_fake_dep_packages` should create the .lock file so that we can test that the source-build script will ignore it.

File: source-container-build/app/test_source_build.py [hunk 4]
```diff
@@ -0,0 +1,1154 @@
...
+        ]
+        self._test_gather_deps_by_package_manager(gomod_deps, ".zip")
```
Comment (chmeliik) [inline source-container-build/app/test_source_build.py:762]: Maybe these 3 should just be one test? I think it would simplify the test

And we want source-build to work when all the package managers are present at the same time, so IMO it makes sense to test it like that

File: source-container-build/app/test_source_build.py [hunk 5]
```diff
@@ -0,0 +1,1154 @@
...
+            "--write-result-to",
+            self.result_file,
+            "--source-dir",
+            self.app_source_dirs.cloned_dir,
+            "--output-binary-image",
+            OUTPUT_BINARY_IMAGE,
+        ]
+        with patch("sys.argv", cli_cmd):
+            with self.assertRaises(SystemExit):
+                source_build.main()
```
Comment (chmeliik) [inline source-container-build/app/test_source_build.py:1154]: We have a lot of tests for failure cases, but they don't assert any specific exception type or error message. Do we even have error handling for these cases? Is it necessary to test them?

File: source-container-build/app/test_source_build.py [hunk 6]
```diff
@@ -0,0 +1,1154 @@
...
+            include_parent_image_sources=True, mock_tarfile_open=tarfile_open
+        )
+
+    @patch("source_build.extract_blob_member")
+    @patch("os.unlink")
+    @patch("tarfile.open")
+    def test_include_parent_image_sources_2(self, tarfile_open, unlink, extract_blob_member):
+        self._test_include_sources(
+            include_parent_image_sources=True,
+            parent_image_with_digest=True,
```
Comment (chmeliik) [inline source-container-build/app/test_source_build.py:1035]: What is this the difference between parent image with digest and without digest? I don't see how `_test_include_sources` checks anything digest-related

File: source-container-build/app/requirements-dev.in [hunk 7]
```diff
@@ -0,0 +1,4 @@
+-r requirements.txt
+
+pytest
+pytest-cov
```
Comment (chmeliik) [inline source-container-build/app/requirements-dev.in:4]: Btw, why do we need pytest if the tests are implemented with the stdlib `unittest`?

Comment (tkdchen) [reply source-container-build/app/requirements-dev.in:4]: It's just a choice of test runners and is bound to the selected test framework used to implement the test. By the way, I like the way with pytest to run tests and output kinds of coverage with a single command and it's much straightforward.

File: source-container-build/app/test_source_build.py (hunk 6 above)
Comment (tkdchen) [reply source-container-build/app/test_source_build.py:1035]: If digest is included in the parent image, it will be removed and the build proceeds. This argument indicates to add digest to the parent image and the rest should run as expected.

File: source-container-build/app/test_source_build.py (hunk 5 above)
Comment (tkdchen) [reply source-container-build/app/test_source_build.py:1154]: Two kinds of failure cases.

One of them happens in the individual method called during the build process, for example `extract_blob_member`. Such methods are not responsible for handling errors, instead, the error will be propagated then be handled by `main`. So, the related tests ensure that failures, the failed subprocess run, are raised as a subprocess error.

Another kind of failure is caused by input, like what your comment is for. It is handled by the `add_argument` `type` argument. For this case, just ensure the command exits by raising `SystemExit`.

File: source-container-build/app/source_build.py (hunk 1 above)
Comment (tkdchen) [reply source-container-build/app/source_build.py]: This change is removed. I'll open another PR for this change.

File: source-container-build/app/test_source_build.py (hunk 4 above)
Comment (tkdchen) [reply source-container-build/app/test_source_build.py]: I'd like to keep these 3 separate cases so that the test shows the specific cases explicitly that the implementation should work.

A new test is added for the the case of "all the package managers are present at the same time".

File: source-container-build/app/requirements-dev.in (hunk 7 above)
Comment (chmeliik) [reply source-container-build/app/requirements-dev.in:4]: Ok, that makes sense. To be honest, I don't even know how to run tests with unittest :smile:

File: source-container-build/app/test_source_build.py (hunk 6 above)
Comment (chmeliik) [reply source-container-build/app/test_source_build.py:1044]: I still don't understand how the test checks digest handling. In fact, I removed these lines and the test still passed :thinking: 

```diff
diff --git a/source-container-build/app/source_build.py b/source-container-build/app/source_build.py
index 870d852..356f5ad 100644
--- a/source-container-build/app/source_build.py
+++ b/source-container-build/app/source_build.py
@@ -426,10 +426,6 @@ def build(args) -> BuildResult:
                 "Multiple base images are specified: %s", ", ".join(args.base_images.splitlines())
             )
         base_image = base_images[-1]
-        if "@sha256:" in base_image:
-            # Remove the digest in case it is included in the image.
-            # e.g. build-container.results.BASE_IMAGES_DIGESTS include it.
-            base_image = base_image.split("@")[0]
         prepared = prepare_base_image_sources(base_image, work_dir, sib_dirs)
         build_result["base_image_source_included"] = prepared
     else:

```

File: source-container-build/app/test_source_build.py [hunk 8]
```diff
@@ -0,0 +1,1162 @@
...
+        expected_src_dirs = [
+            os.path.join(prefetched_sources_dir, f"src-{count}")
+            for count in range(len(deps_with_known_file_ext))
+        ]
+        self.assertListEqual(sorted(expected_src_dirs), sorted(sib_dirs.extra_src_dirs))
+
+        gathered_deps = []  # collect the dep packages gathered by the method
+        for dir_path, subdir_names, file_names in os.walk(prefetched_sources_dir):
+            for name in file_names:
+                gathered_deps.extend([name for ext in dep_exts if name.endswith(ext)])
```
Comment (chmeliik) [inline source-container-build/app/test_source_build.py:741]: Why does the test filter both the expected data (`deps_with_known_file_ext`) and the actual data (`gathered_deps`) by file extension?

If the script added unwanted files to the `prefetched_sources_dir`, I think this test wouldn't notice. IMO we shouldn't filter the actual data, we should add all the `file_names` to `gathered_deps`

File: source-container-build/app/test_source_build.py (hunk 5 above)
Comment (chmeliik) [reply source-container-build/app/test_source_build.py:1162]: Ok, seems reasonable

**Comment by chmeliik** [comment]
Two more comments, otherwise LGTM

File: source-container-build/app/test_source_build.py (hunk 6 above)
Comment (tkdchen) [reply source-container-build/app/test_source_build.py:1048]: Missing assertion for removing digest. Added :)

File: source-container-build/app/test_source_build.py (hunk 8 above)
Comment (tkdchen) [reply source-container-build/app/test_source_build.py]: You are right. Updated.

**Review by mkosiarc — APPROVED** [review]

**Comment by tkdchen** [comment]
Thanks all for your review. Merging.
